File: app/src/main/java/jp/hazuki/yuzubrowser/adblock/fragment/AdBlockEditDialog.java
/*
 * Copyright (C) 2017 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.adblock.fragment;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.text.InputType;
import android.text.TextUtils;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.LinearLayout;

import jp.hazuki.yuzubrowser.adblock.AdBlock;
import jp.hazuki.yuzubrowser.utils.extensions.ContextExtensionsKt;

public class AdBlockEditDialog extends DialogFragment {
    private static final String ARG_TITLE = "title";
    private static final String ARG_INDEX = "index";
    private static final String ARG_ID = "id";
    private static final String ARG_TEXT = "text";

    private AdBlockEditDialogListener listener;

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        final EditText editText = new EditText(getActivity());
        final LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT);
        float density = ContextExtensionsKt.getDensity(getActivity());
        int marginWidth = (int) (4 * density + 0.5f);
        int marginHeight = (int) (16 * density + 0.5f);
        layoutParams.setMargins(marginWidth, marginHeight, marginWidth, marginHeight);
        editText.setLayoutParams(layoutParams);
        editText.setId(android.R.id.edit);
        editText.setInputType(InputType.TYPE_CLASS_TEXT);

        String text = getArguments().getString(ARG_TEXT);
        if (!TextUtils.isEmpty(text))
            editText.setText(text);

        return new AlertDialog.Builder(getActivity())
                .setView(editText)
                .setTitle(getArguments().getString(ARG_TITLE))
                .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        listener.onEdited(
                                getArguments().getInt(ARG_INDEX, -1),
                                getArguments().getInt(ARG_ID, -1),
                                editText.getText().toString());

                    }
                })
                .setNegativeButton(android.R.string.cancel, null)
                .create();
    }

    public static AdBlockEditDialog newInstance(String title) {
        return newInstance(title, -1, null);
    }

    public static AdBlockEditDialog newInstance(String title, int index, AdBlock adBlock) {
        AdBlockEditDialog dialog = new AdBlockEditDialog();
        Bundle bundle = new Bundle();
        bundle.putString(ARG_TITLE, title);
        bundle.putInt(ARG_INDEX, index);
        if (adBlock != null) {
            bundle.putInt(ARG_ID, adBlock.getId());
            bundle.putString(ARG_TEXT, adBlock.getMatch());
        }
        dialog.setArguments(bundle);
        return dialog;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (getParentFragment() instanceof AdBlockEditDialogListener) {
            listener = (AdBlockEditDialogListener) getParentFragment();
        } else {
            listener = (AdBlockEditDialogListener) getActivity();
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        listener = null;
    }

    interface AdBlockEditDialogListener {
        void onEdited(int index, int id, String text);
    }
}

File: module/adblock/src/main/java/jp/hazuki/yuzubrowser/adblock/ui/original/AdBlockEditDialog.kt
/*
 * Copyright (C) 2017-2019 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.adblock.ui.original

import android.app.AlertDialog
import android.app.Dialog
import android.content.Context
import android.os.Bundle
import android.text.InputType
import android.view.ViewGroup
import android.widget.EditText
import android.widget.LinearLayout
import androidx.fragment.app.DialogFragment
import jp.hazuki.yuzubrowser.adblock.repository.original.AdBlock
import jp.hazuki.yuzubrowser.core.utility.extensions.density

class AdBlockEditDialog : DialogFragment() {

    private var listener: AdBlockEditDialogListener? = null

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {

        val params = LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT)
        val density = activity!!.density
        val marginWidth = (4 * density + 0.5f).toInt()
        val marginHeight = (16 * density + 0.5f).toInt()
        params.setMargins(marginWidth, marginHeight, marginWidth, marginHeight)
        val editText = EditText(activity).apply {
            layoutParams = params
            id = android.R.id.edit
            inputType = InputType.TYPE_CLASS_TEXT
        }

        val arguments = arguments ?: throw NullPointerException()
        val text = arguments.getString(ARG_TEXT)
        if (!text.isNullOrEmpty())
            editText.setText(text)

        return AlertDialog.Builder(activity)
                .setView(editText)
                .setTitle(arguments.getString(ARG_TITLE))
                .setPositiveButton(android.R.string.ok) { _, _ ->
                    listener!!.onEdited(
                            arguments.getInt(ARG_INDEX, -1),
                            arguments.getInt(ARG_ID, -1),
                            editText.text.toString())
                }
                .setNegativeButton(android.R.string.cancel, null)
                .create()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        listener = if (parentFragment is AdBlockEditDialogListener) {
            parentFragment as AdBlockEditDialogListener
        } else {
            activity as AdBlockEditDialogListener
        }
    }

    override fun onDetach() {
        super.onDetach()
        listener = null
    }

    internal interface AdBlockEditDialogListener {
        fun onEdited(index: Int, id: Int, text: String)
    }

    companion object {
        private const val ARG_TITLE = "title"
        private const val ARG_INDEX = "index"
        private const val ARG_ID = "id"
        private const val ARG_TEXT = "text"

        operator fun invoke(title: String, index: Int = -1, adBlock: AdBlock? = null): AdBlockEditDialog {
            return AdBlockEditDialog().apply {
                arguments = Bundle().apply {
                    putString(ARG_TITLE, title)
                    putInt(ARG_INDEX, index)
                    if (adBlock != null) {
                        putInt(ARG_ID, adBlock.id)
                        putString(ARG_TEXT, adBlock.match)
                    }
                }
            }
        }
    }
}
